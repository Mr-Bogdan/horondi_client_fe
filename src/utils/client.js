import { ApolloClient, gql } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';
import fetch from 'unfetch';
import { InMemoryCache, IntrospectionFragmentMatcher } from 'apollo-cache-inmemory';
import { createUploadLink } from 'apollo-upload-client/public';

import { USER_TOKENS, FETCH_POLICY, USER_IS_BLOCKED } from '../configs';
import { AUTH_ERRORS } from '../const/error-messages';
import { getFromLocalStorage } from '../services/local-storage.service';
import refreshAuthToken from './regenerateAuthTokenPair';

const introspectionResult = require('../fragmentTypes');

const { ACCESS_TOKEN } = USER_TOKENS;
const fragmentMatcher = new IntrospectionFragmentMatcher({
  introspectionQueryResultData: introspectionResult
});

export const REACT_APP_API_URL =
  window.env && window.env.REACT_APP_API_URL
    ? window.env.REACT_APP_API_URL
    : process.env.REACT_APP_API_URL;

const authLink = setContext((_, { headers }) => {
  const token = getFromLocalStorage(ACCESS_TOKEN);
  return {
    headers: {
      ...headers,
      token: token || ''
    }
  };
});

export const client = new ApolloClient({
  fetch,
  link: authLink.concat(createUploadLink({ uri: REACT_APP_API_URL })),
  cache: new InMemoryCache({
    addTypename: false,
    fragmentMatcher
  })
});

export const getItems = async (query, variables = {}) => {
  try {
    const token = getFromLocalStorage(ACCESS_TOKEN);

    const queryResult = await client.query({
      query: gql`
        ${query}
      `,
      variables,
      context: {
        headers: {
          token
        }
      },
      fetchPolicy: FETCH_POLICY
    });

    if (
      queryResult.data &&
      Object.values(queryResult.data)[0]?.message === AUTH_ERRORS.ACCESS_TOKEN_IS_NOT_VALID
    ) {
      const tokenResult = await refreshAuthToken();

      if (tokenResult) {
        return await getItems(query, variables);
      }
      throw new Error(AUTH_ERRORS.REFRESH_TOKEN_IS_NOT_VALID);
    } else if (
      queryResult.data &&
      Object.values(queryResult.data)[0]?.message === USER_IS_BLOCKED
    ) {
      throw new Error(USER_IS_BLOCKED);
    } else {
      return queryResult;
    }
  } catch (e) {
    throw new Error(e.message);
  }
};

export const setItems = async (query, variables) => {
  try {
    const token = getFromLocalStorage(ACCESS_TOKEN);

    const mutationResult = await client.mutate({
      mutation: gql`
        ${query}
      `,
      variables,
      context: {
        headers: {
          token
        }
      },
      fetchPolicy: FETCH_POLICY
    });

    if (
      mutationResult.data &&
      Object.values(mutationResult.data)[0]?.message === AUTH_ERRORS.ACCESS_TOKEN_IS_NOT_VALID
    ) {
      const tokenResult = await refreshAuthToken();

      if (tokenResult) {
        return await setItems(query, variables);
      }
      throw new Error(AUTH_ERRORS.REFRESH_TOKEN_IS_NOT_VALID);
    } else if (
      mutationResult.data &&
      Object.values(mutationResult.data)[0]?.message === USER_IS_BLOCKED
    ) {
      throw new Error(USER_IS_BLOCKED);
    } else {
      return mutationResult;
    }
  } catch (e) {
    throw new Error(e.message);
  }
};

export const setChatMail = (query, variables) =>
  client.mutate({
    mutation: gql`
      ${query}
    `,
    variables
  });
