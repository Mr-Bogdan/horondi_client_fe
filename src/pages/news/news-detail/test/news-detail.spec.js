import React from 'react';
import { useQuery } from '@apollo/client';
import NewsDetail from '../news-detail';

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: () => null, i18n: { language: 'ua' } })
}));
jest.mock('../../news-detail/news-detail.style', () => ({ useStyles: () => ({}) }));
jest.mock('@apollo/client');
jest.mock('react-redux');

let wrapper;
const useQueryData = {
  loading: false,
  error: false,
  data: {}
};

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: () => [{ _id: [0], text: [''], title: [''], author: { name: '' } }, () => null]
}));

describe('test newsDetail', () => {
  it('should match snapshot', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData
    }));
    wrapper = shallow(<NewsDetail match={{ params: { id: '' } }} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should cover other branches', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData,
      loading: true
    }));

    wrapper = shallow(<NewsDetail match={{ params: { id: '' } }} />);
  });

  it('should cover rest branches', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData,
      error: {}
    }));

    wrapper = shallow(<NewsDetail match={{ params: { id: '' } }} />);
  });
});
