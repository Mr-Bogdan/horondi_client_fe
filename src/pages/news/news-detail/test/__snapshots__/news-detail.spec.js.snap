// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test newsDetail should match snapshot 1`] = `
<WithStyles(ForwardRef(Card))>
  <WithStyles(ForwardRef(CardContent))>
    <WithStyles(ForwardRef(Typography))
      component="h2"
      gutterBottom={true}
      variant="h2"
    />
    <WithStyles(ForwardRef(CardHeader))
      subheader="Invalid Date"
    />
    <hr />
    <div>
      <WithStyles(ForwardRef(CardMedia))
        alt={null}
        component="div"
        image="https://horondi.blob.core.windows.net/horondi/images/undefined"
        title={null}
      />
    </div>
    <WithStyles(ForwardRef(Typography))
      color="textSecondary"
      component="div"
      id="fullText"
      variant="body2"
    />
    <hr />
    <div>
      <WithStyles(ForwardRef(CardHeader))
        id="newsAuthor"
        subheader={null}
      />
      <WithStyles(ForwardRef(CardMedia))
        component="div"
        id="newsAuthorAvatar"
        image="https://horondi.blob.core.windows.net/horondi/images/undefined"
        title={null}
      />
    </div>
  </WithStyles(ForwardRef(CardContent))>
</WithStyles(ForwardRef(Card))>
`;
